README.md
# PYRMM: (Learned) Risk Metric Maps in Python

## Overview

![Dubins Example in 2D](docs/20220915_DubinsRiskMetricResults_2.png)

The pyrmm library is used for development of neural network models that predict risk (e.g. probability of failure) for kinodynamic systems (e.g. cars, airplanes, etc.)

## Citation

TODO

### Reproduction of Paper Results

Please see instructions in the [`experiments` sub-package](https://github.com/mit-drl/pyrmm/tree/master/experiments#readme)

## Disclaimer

Research was sponsored by the United States Air Force Research Laboratory and the United States Air Force Artificial Intelligence Accelerator and was accomplished under Cooperative Agreement Number FA8750-19-2-1000. The views and conclusions contained in this document are those of the authors and should not be interpreted as representing the official policies, either expressed or implied, of the United States ​Air Force or the U.S. Government. The U.S. Government is authorized to reproduce and distribute reprints for Government purposes notwithstanding any copyright notation herein.

## Installation

Due to several 3rd party non-Python dependencies, the install process is somewhat involved and manual.

The installation process is written for an Ubuntu Linux machine.

It is __strongly__ recommended to use a conda environment to manage the installation. These instructions only cover a conda environment installation process.

### Clone and Conda Env

```bash
git clone --recurse-submodules git@github.com:mit-drl/pyrmm.git
cd pyrmm

# creates a conda env called pyrmm, this may take a while
conda env create -f environment.yml 
conda activate pyrmm

# build the mazegenerator submodule used for procedurally
# generating 2D maze images
cd mazegenerator/src
make

# build the V-HACD library used to create convex decompositions of procedurally generated 
# rooms
cd ../v-hacd/app/
cmake -DCMAKE_BUILD_TYPE=Release CMakeLists.txt
cmake --build .
```

### Install OMPL with Python Bindings

Next we will download and install OMPL following the [OMPL install instructions](https://ompl.kavrakilab.org/core/installation.html). 

We use the [Open Motion Planning Library (OMPL)](https://ompl.kavrakilab.org/core/index.html) for it's state and control space definitions and sampling functions. OMPL is a C++ library but provides python bindings that must be built by running an installation script. These instructions were written when [ompl 1.5](https://github.com/ompl/ompl/releases/tag/1.5.0) was the most recent release, but now [ompl 1.6](https://github.com/ompl/ompl/releases). The install process *should* be the same for ompl 1.6, but it hasn't been tested.

```bash
# Download OS-specific script from https://ompl.kavrakilab.org/core/installation.html
# For ubuntu, the next steps look like
chmod u+x install-ompl-ubuntu.sh
./install-ompl-ubuntu.sh --python # this will install the latest release of OMPL with Python bindings
```

> :warning: **Troubleshooting**
> The python bindings can take __hours__ to build. It is [recommended to use PyPy3](https://github.com/ompl/ompl/issues/699#issuecomment-961580319) to speed up the process, but these instructions don't go into the level of detail needed to walkthrough PyPy3 vs CPython

The above steps should install ompl to your home directory, e.g. `/home/user/ompl-1.5.2/`. The python bindings are not immediately "discoverable" by python; we need to point the conda environment to the ompl install using a `.pth` file. Assuming your conda environment is stored in `~/miniconda3/envs/pyrmm`, you run something similar to (adjusting paths as need to match install locations and username on your machine)

```bash
touch ~/miniconda3/envs/pyrmm/lib/python3.8/site-packages/ompl.pth
echo "/home/user/ompl-1.5.2/py-bindings" >> ~/miniconda3/envs/pyrmm/lib/python3.8/site-packages/ompl.pth
```

### Install Blender


We use [blender](https://www.blender.org/) in order to convert `.stl` meshes produced by [`pcg-gazebo`](https://github.com/boschresearch/pcg_gazebo/) into `.obj` files that can be loaded by [pybullet](https://pybullet.org/wordpress/). These instructions were written with blender 3.2.1, but the pyrmm library *should* be compatible with later version of blender

Download and install blender following instructions here: https://www.blender.org/download/

To make blender more easily callable without the entire path to the install, create the following script (assuming blender 3.2.1 was installed to home directory) and save to `/usr/local/bin/blender`

```bash
#!/bin/sh
~/blender/blender-3.2.1-linux-x64/blender "$@"
```

Then make that script executable by running 

```bash
sudo chmod a+x /usr/local/bin/blender
```

### Verify Installation

TODO

## Example Usage: Dubins Vehicle in PPM Mazes

### Procedural Generation of Obstacle Space

TODO

### Approximate Risk Metric Data Generation

TODO

### Risk Metric Model Training

TODO

## Example Usage: Parallel Autonomy for Dubins Vehicle

TODO


CHANGELOG.md
# Change Log

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [v0.0.5] - 2023-03-13

**Major Refactor**

+ Added first implementation and tests of Inferred Risk Barrier Functions (IRBF, although currently referred to in code as "CBFLRMM" and "RiskCBF")
+ Abstracted key SystemSetup functions out of child-class parent class, e.g. `sample_control_numpy`, `propagate_path`, `path_ompl_to_numpy`, and `path_numpy_to_ompl`
+ Refactoring all system setups to have conversion functions `state_ompl_to_numpy`, `state_numpy_to_ompl`, `control_ompl_to_numpy`, `control_numpy_to_ompl`
+ Completely eliminated `StatePropagator` classes in favor of a single `propagate_path` function

## [v0.0.4] - 2022-09-14

Major update. complete implementation of parallel autonomy case study with dubins-4d vehicle compared to CBF and HJ-reach agents

notebooks/irbf_example.py
# Example of a risk-metric control barrier function
# for 1D Double integrator System

import torch
import numpy as np
import matplotlib.pyplot as plt
import itertools

from functools import partial

import pyrmm.utils.utils as U
from pyrmm.environments.dubins4d_reachavoid import cvx_qp_solver

from pyrmm.setups.double_integrator import DoubleIntegrator1DSetup

from pyrmm.modelgen.data_modules import LSFORDataModule
from pyrmm.modelgen.modules import ShallowRiskCBFPerceptron, CBFLRMMModule, DeepRiskCBFPerceptron
from pyrmm.modelgen.double_integrator import local_states_datagen, \
    trivial_state_feature_map, first_order_state_feature_map, quadratic_state_feature_map,  cubic_state_feature_map

###
# Problem parameters
###

# obstacle location on x-axis
P_OBS = 5.0 # [m]

# control bounds
U1_MIN = -1.0   # [m/s/s]
U1_MAX = 1.0    # [m/s/s]

# evaluate derivatives at current state, i.e. local origin
XTIL_1 = 0.0
XTIL_2 = 0.0

# initial/current state of system, or 
# state at which observation z is taken
P_Z_ARR = np.linspace(-1.0, 4.5)    # [m]
# V_Z_ARR = np.arange(-1, 2.5, 0.5)   # [m/s]
V_Z_ARR = np.linspace(-1, 2.5)   # [m/s]

# plotting params
MARKER = itertools.cycle((',', '+', '.', 'o', '*')) 

def run_analytical_cbf_analysis():

    results = np.empty((len(V_Z_ARR), len(P_Z_ARR)))

    for i, v_z in enumerate(V_Z_ARR):
        for j, p_z in enumerate(P_Z_ARR):

            ###
            # Computed parameters
            ###

            # control input for minimum stopping distance
            if v_z >= 0:
                u1_minstop = U1_MIN
            else:
                u1_minstop = U1_MAX

            # compute final position of minimum-distance stopping
            p_minstop = -0.5 * v_z**2 / u1_minstop + p_z

            # break if stopping impossible
            if p_minstop > P_OBS:
                results[i,j] = np.nan
                continue

            # risk metric at current state
            rho_x = np.exp(-(P_OBS - p_minstop))

            # barrier function value at current state
            h_x = 1 - rho_x

            ###
            # Compute Lie derivatives
            ###

            # Lie derivative along g(x)
            Lghx1 = ((XTIL_2 + v_z) / u1_minstop) * np.exp(p_minstop - P_OBS)

            # Lie derivative along f(x)
            Lfhx = -(XTIL_2 + v_z) * np.exp(p_minstop - P_OBS)

            ### 
            # Formulate QP
            ###

            # objective: minimize control input
            P = np.eye(1).reshape(1,1)
            q = np.zeros(1).reshape(1,1)
            lambda_h = 0.1

            # Constraint: Left side of <= inequality
            G = np.reshape([-Lghx1], (1,1))

            # Constraint: right side of <= inequality
            h = np.reshape([Lfhx + lambda_h * h_x], (1,1))

            ### 
            # solve QP
            ###

            u_opt = cvx_qp_solver(P=P, q=q, G=G, h=h)
            results[i,j] = u_opt[0]

            # print("Min-norm safe control = ", u_opt[0])

        plt.plot(P_Z_ARR, results[i], marker = next(MARKER), label="v={} m/s".format(v_z))

    plt.title("Safe Control Input As Function of Position and Velocity")
    plt.xlabel("position [m]")
    plt.ylabel("control input [m/s/s]")
    plt.legend()
    plt.show()

# create model object to load saved model into
# Note: this is hacky to have to remember the exact number of observation
# inputs, state features, and neurons. There has got to be a better way 
# save these params at model save time
n_obsv_dim = 3
# n_feat_dim = 4
n_feat_dim = 6 
# n_feat_dim = 8 
# n_feat_dim = 2
# n_neurons = 8
# rmcbf_model = ShallowRiskCBFPerceptron(
#     num_obs_inputs=n_obsv_dim,
#     num_state_features=n_feat_dim,
#     num_neurons=n_neurons
# )
rmcbf_model = DeepRiskCBFPerceptron(
    num_obs_inputs=n_obsv_dim,
    num_state_features=n_feat_dim,
    num_neurons=[32, 16, 8]
)

# load checkpoint
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     "outputs/2023-01-19/12-34-03/lightning_logs/version_0/" +
#     "checkpoints/epoch=511-step=308735.ckpt"
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     "outputs/2023-01-20/15-35-25/lightning_logs/version_0/" +
#     "checkpoints/epoch=511-step=308735.ckpt"
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     "outputs/2023-02-03/13-43-20/lightning_logs/version_0/checkpoints/epoch=511-step=308735.ckpt"
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-07/08-08-31/lightning_logs/version_0/checkpoints/epoch=284-step=2125529.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-07/12-49-53/lightning_logs/version_0/epoch=71-step=1339919.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-07/15-02-24/lightning_logs/version_0/epoch=263-step=1240535.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-07/16-39-56/lightning_logs/version_0/checkpoints/epoch=247-step=1111287.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-07/19-08-10/lightning_logs/version_0/epoch=123-step=555643.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-07/20-59-56/lightning_logs/version_0/epoch=66-step=300226.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-07/21-36-53/lightning_logs/version_0/epoch=73-step=347725.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-07/22-48-57/lightning_logs/version_0/epoch=118-step=501822.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-08/07-24-31/lightning_logs/version_0/epoch=41-step=781619.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-08/09-24-03/lightning_logs/version_0/epoch=42-step=800229.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-08/12-07-19/lightning_logs/version_0/epoch=10-step=204709.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-08/12-07-19/lightning_logs/version_0/epoch=52-step=986329.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-08/15-14-06/lightning_logs/version_0/epoch=28-step=1078103.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-08/18-10-51/lightning_logs/version_0/epoch=165-step=780033.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-09/17-00-05/lightning_logs/version_0/checkpoints/saved_epoch=87-step=1637679.ckpt'
# )
# chkpt_file = (
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     'outputs/2023-02-09/17-00-05/lightning_logs/version_0/checkpoints/saved_epoch=389-step=7257899.ckpt'
# )
chkpt_file = (
    "/home/ross/Projects/AIIA/risk_metric_maps/" +
    'outputs/2023-02-15/11-06-29/lightning_logs/version_0/checkpoints/epoch=511-step=2158079.ckpt'
)
chkpt = torch.load(chkpt_file)

# create pytorch lightning module
# no optimizer because this is only for eval/inference
rmcbf = CBFLRMMModule(num_inputs=3, model=rmcbf_model, optimizer=None)

# load checkpoint into module
rmcbf.load_state_dict(chkpt['state_dict'])
rmcbf.eval()

# create data module and load training data to get observation scaler
# datapaths = U.get_abs_pt_data_paths(
#     "/home/ross/Projects/AIIA/risk_metric_maps/" +
#     "outputs/2022-12-19/14-09-56/"
# )
datapaths = U.get_abs_pt_data_paths(
    "/home/ross/Projects/AIIA/risk_metric_maps/" +
    "outputs/2023-02-03/15-30-08/"
)

# local_states_datagen_func = partial(local_states_datagen, 1.5)  # again, this is just something you need to know was used during modelgen
# local_states_datagen_func = partial(local_states_datagen, 3)  # again, this is just something you need to know was used during modelgen
local_states_datagen_func = partial(local_states_datagen, 1.0)  # again, this is just something you need to know was used during modelgen
rmcbf_data_mod = LSFORDataModule(
    datapaths=datapaths,
    val_ratio=0,
    batch_size=1,
    num_workers=0,
    # state_feature_map=first_order_state_feature_map,  # again this is something you just need to know was used during modelgen, very hacky/brittle to encode this way
    state_feature_map=quadratic_state_feature_map,  # again this is something you just need to know was used during modelgen, very hacky/brittle to encode this way
    # state_feature_map=trivial_state_feature_map,  # again this is something you just need to know was used during modelgen, very hacky/brittle to encode this way
    # state_feature_map=cubic_state_feature_map,  # again this is something you just need to know was used during modelgen, very hacky/brittle to encode this way
    local_states_datagen=local_states_datagen_func,
    compile_verify_func=None
)
rmcbf_data_mod.setup("test")

# create double integrator setup to query proper observations
pos_bounds = [-32, 32]
vel_bounds = [-8, 8]
acc_bounds = [U1_MIN, U1_MAX] # deterministic system with no accel
obst_bounds = [P_OBS, P_OBS+0.1]    # very slim obstacle
di1d_setup = DoubleIntegrator1DSetup(
    pos_bounds=pos_bounds, 
    vel_bounds=vel_bounds, 
    acc_bounds=acc_bounds, 
    obst_bounds=obst_bounds)

def run_irbf_global_analysis():

    # iterate through states
    risks = np.empty((len(V_Z_ARR), len(P_Z_ARR)))
    cbf_weights = np.empty((len(V_Z_ARR), len(P_Z_ARR), n_feat_dim))
    ctrls = np.empty((len(V_Z_ARR), len(P_Z_ARR)))
    for i, v_z in enumerate(V_Z_ARR):
        for j, p_z in enumerate(P_Z_ARR):

            print("DEBUG: i,j = ",i,j)

            # package state in np array and convert state to ompl
            s_z_np = np.array([p_z, v_z])
            s_z_ompl = di1d_setup.space_info.allocState()
            di1d_setup.state_numpy_to_ompl(npState=s_z_np, omplState=s_z_ompl)
            print("DEBUG: state = ",s_z_np)

            # get observation of state
            o_z_np = di1d_setup.observeState(state=s_z_ompl)
            print("DEBUG: observation = ",o_z_np)

            # scale observation
            o_z_scaled_pt = torch.from_numpy(
                np.float32(
                    rmcbf_data_mod.observation_scaler.transform(
                        o_z_np.reshape(1,n_obsv_dim)
                    )
                )
            ).reshape(n_obsv_dim,)
            print("DEBUG: scaled observation = ",o_z_scaled_pt)

            # compute state features of localized state
            stil_z_np = np.array([XTIL_1, XTIL_2])
            ftil_z_np = rmcbf_data_mod.state_feature_map(stil_z_np)
            print("DEBUG: state feature of local state = ", ftil_z_np)

            # infer risk metric and output cbf weights from trained model
            rho_z_pt, w_cbf_pt = rmcbf.forward(observation=o_z_scaled_pt, state_features=torch.from_numpy(ftil_z_np))
            risks[i,j] = rho_z_pt.detach().numpy()
            cbf_weights[i,j,:] = w_cbf_pt.detach().numpy()
            print("DEBUG: cbf weights = ", cbf_weights[i,j,:])
            print("DEBUG: inferred risk metric = ", rho_z_pt)

        plt.plot(P_Z_ARR, risks[i], marker = next(MARKER), label="v={} m/s".format(v_z))

    plt.title("Risk Estimate As Function of Position and Velocity\n1-D Double Integrator System with Obstacle at position=5")
    plt.xlabel("position [m]")
    plt.ylabel("probability of failure [-]")
    plt.legend()
    plt.show()


    p_mesh, v_mesh = np.meshgrid(P_Z_ARR, V_Z_ARR)
    fig, ax = plt.subplots()
    pcm = ax.pcolormesh(p_mesh, v_mesh, risks, cmap="turbo", vmin=0, vmax=1)
    fig.colorbar(pcm, ax=ax)
    plt.title("Risk Estimate (Failure Probability) Inferred at Each State\n"+
        "for 1-D Double Integrator System with Uniform Random Policy and\n"+
        "Obstacle at position=5")
    plt.xlabel("position [m]")
    plt.ylabel("velocity [m/s]")
    plt.show()

    #################################

def run_irbf_local_analysis():

    # analyze inferred risk function from single observation
    # package state in np array and convert state to ompl
    # s_z_np = np.zeros(2)
    # s_z_np = np.array([4.5, 0.0])
    # s_z_np = np.array([4.5, 2.0])
    # s_z_np = np.array([2.0, 1.5])
    # s_z_np = np.array([3.0, 1.0])
    # s_z_np = np.array([3.2, 1.2])
    s_z_np = np.array([1.0, 0.5])
    s_z_ompl = di1d_setup.space_info.allocState()
    di1d_setup.state_numpy_to_ompl(npState=s_z_np, omplState=s_z_ompl)
    print("DEBUG: state = ",s_z_np)

    # get observation of state
    o_z_np = di1d_setup.observeState(state=s_z_ompl)
    print("DEBUG: observation = ",o_z_np)

    # scale observation
    o_z_scaled_pt = torch.from_numpy(
        np.float32(
            rmcbf_data_mod.observation_scaler.transform(
                o_z_np.reshape(1,n_obsv_dim)
            )
        )
    ).reshape(n_obsv_dim,)
    print("DEBUG: scaled observation = ",o_z_scaled_pt)

    # iterate through states
    risks_0_0 = np.empty((len(V_Z_ARR), len(P_Z_ARR)))
    # stil_z_np = np.empty((len(V_Z_ARR), len(P_Z_ARR), 2))
    for i, v_z in enumerate(V_Z_ARR):
        for j, p_z in enumerate(P_Z_ARR):

            # compute state features of localized state
            stil_z_np = np.array([p_z, v_z]) - s_z_np
            ftil_z_np = rmcbf_data_mod.state_feature_map(stil_z_np)
            print("DEBUG: state feature of local state = ", ftil_z_np)

            # infer risk metric and output cbf weights from trained model
            rho_z_pt, w_cbf_pt = rmcbf.forward(observation=o_z_scaled_pt, state_features=torch.from_numpy(ftil_z_np))
            risks_0_0[i,j] = rho_z_pt.detach().numpy()
            # cbf_weights[i,j,:] = w_cbf_pt.detach().numpy()
            # print("DEBUG: cbf weights = ", cbf_weights[i,j,:])
            print("DEBUG: cbf weights = ", w_cbf_pt.detach().numpy())

        # plt.plot(P_Z_ARR, risks_0_0[i], marker = next(MARKER), label="v={} m/s".format(v_z))

    # ref: https://stackoverflow.com/a/54088910/4055705
    p_mesh, v_mesh = np.meshgrid(P_Z_ARR, V_Z_ARR)
    fig, ax = plt.subplots()
    pcm = ax.pcolormesh(p_mesh, v_mesh, risks_0_0, cmap="turbo", vmin=0, vmax=1)
    fig.colorbar(pcm, ax=ax)
    plt.plot(s_z_np[0], s_z_np[1], 'k+', ms=10, mew=3)
    plt.title("Risk Estimate (Failure Probability) Extrapolated from\n"+
        "Single-Point Risk Function at p={} [m], v= {} [m/s]\n".format(s_z_np[0], s_z_np[1])+
        "for 1-D Double Integrator System with\n" + 
        "Uniform Random Policy and Obstacle at position=5")
    plt.xlabel("position [m]")
    plt.ylabel("velocity [m/s]")
    plt.show()

    # plt.title("Risk Estimate As Function of Position and Velocity\nfrom Single Point Risk Function\n1-D Double Integrator System with Obstacle at position=5")
    # plt.xlabel("position [m]")
    # plt.ylabel("probability of failure [-]")
    # plt.legend()
    # plt.show()


if __name__ == "__main__":
    # run_analytical_cbf_analysis()
    run_irbf_global_analysis()
    run_irbf_local_analysis()
